presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java

```diff
@@ -77,7 +77,7 @@ public IcebergResourceFactory(IcebergConfig config, IcebergCatalogName catalogNa
     public Catalog getCatalog(ConnectorSession session)
     {
         try {
-            return catalogCache.get(getCatalogCacheKey(session), () -> CatalogUtil.loadCatalog(
+            return catalogCache.get(catalogName, () -> CatalogUtil.loadCatalog(
```

Comment:
@agrawalreetika How can it make sure that a catalog A that is not supposed to be visible to a session remains invisible to that session? If another session has the access to catalog A and the cache caches it, would the first session retrieve it and see things it's not supposed to see?

Reply:
Hi @yingsu00, 
- About already used `User:<user-name>,<Principle:>,Role:` as part of Catalog cache key, I don't see this information is not getting passed to Iceberg library from Presto and is maintained as cache key on Presto side.

- As discussed about these catalog level username & password [here](https://github.com/prestodb/presto/blob/master/presto-iceberg/src/main/java/com/facebook/presto/iceberg/nessie/NessieConfig.java#L78), These are suppose to be used for basic auth of Nessie but while looking at the Nessie client library it looks like BASIC AUTH is deprecated [here](https://github.com/projectnessie/nessie/blob/nessie-0.59.0/api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java#L34) 

So as per this, we can also remove BASIC authentication option for Nessie from Presto side as well.

Reply:
Taking a step back, why did we need to make this change in the first place?

Reply:
@tdcmeehan Since current key is doing below - 
1. On Presto side, Under IcebergResourceFactory there is one more cache present called [catalogCache](https://github.com/prestodb/presto/blob/master/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java#L53), it was introduced as part of initial iceberg connector shipping to Presto.
But it has key as `User:<user-name>,<Principle:>,Role: `and Iceberg Catalog as Values which is causing FileIO Object to get created for every new user & catalog file because of [CatalogUtil.loadCatalog()](https://github.com/apache/iceberg/blob/apache-iceberg-1.3.1/core/src/main/java/org/apache/iceberg/CatalogUtil.java#L216) while loading cache. Current catalogCache is controlled by [iceberg.catalog.cached-catalog-num](https://github.com/prestodb/presto/blob/master/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java#L111) config in Presto.
2. Now [Manifest file caching](https://github.com/apache/iceberg/blob/apache-iceberg-1.3.1/core/src/main/java/org/apache/iceberg/ManifestFiles.java#L68) has FILEIO as key for the cache, can store upto 8 FILEIO Object by default [here](https://github.com/apache/iceberg/blob/apache-iceberg-1.3.1/core/src/main/java/org/apache/iceberg/SystemProperties.java#L40), this will fill that cache if 8 unique user query/ catalog query comes in from Presto.

This is why I think we can change the catalog cache key to `catalog name + catalog session properties` so that we can reuse the same catalog object and avoid loading the Catalog for every unique user & catalog. 
Please let me know if you have any questions or any other suggestions here.

Reply:
@tdcmeehan 
For the catalog cache, the original key was like "User:username,Principle:principle,Role:role,ExtraCredentialKey1:key1,...", and the value was the catalog object. There were several issues here
1. Since the key uniquely identify a <user, principle, role> tuple, it assumes that each tuple maps to ONE catalog. This doesn't seem correct to me. A user can have multiple Iceberg catalogs.
2. Multiple users may have the same <principle, role> tuple, so the other user won't be able retrieve the catalog from the cache even though he/she has the same authentication level.
3. Adding the "User:", "Principle:" string headers to the key is not good for performance. We want to only hash necessary information. The longer the string, the slower the hashing and probing will be.

Actually, the check of access to catalogs were done in `SystemAccessControl`, which was done on the metadata level and happened earlier than this `loadCatalog()`. It seems to me having the principle and role in the catalog cache is unnecessary. If a <user, principle, role> tries to access a catalog he/she is not supposed to see, SystemAccessControl would deny it before getting to here.

cc @ChunxuTang

Reply:
Makes sense!